common/src/block_interval.rs

```diff
@@ -0,0 +1,272 @@
+        // because both separators start with `..`
+        if let Ok(range) = parsing::parse_range_inclusive(s) {
+            Ok(BlockInterval::Range(range))
+        } else if let Ok(range) = parsing::parse_range_exclusive(s) {
+            Ok(BlockInterval::Range(range))
+        }
+        // Now we look for the follow from range
+        else if s.contains("..") {
+            let mut split = s.trim().split("..").filter(|s| *s != "=" && !s.is_empty());
+
+            // Any other character after `..` or `..=` is invalid
+            if split.clone().count() > 1 {
+                return Err(anyhow!("invalid block interval range '{s}'"));
+            }
+            let num = split
+                .next()
+                .map(|num| {
+                    num.parse::<u64>()
+                        .map_err(|_| anyhow!("invalid block number '{num}'"))
+                })
+                .ok_or(anyhow!("invalid block interval range '{s}'"))??;
+            return Ok(BlockInterval::FollowFrom {
+                start_block: num,
+                block_time: None,
+            });
+        }
+        // Only single block number is left to try to parse
+        else {
+            let num: u64 = s
+                .trim()
+                .parse()
+                .map_err(|_| anyhow!("invalid block interval range '{s}'"))?;
+            return Ok(BlockInterval::SingleBlockId(BlockId::Number(num.into())));
+        }
+    }
+
+    /// Convert the block interval into an async stream of block numbers.
+    pub fn into_bounded_stream(self) -> anyhow::Result<impl Stream<Item = u64>> {
+        match self {
+            BlockInterval::SingleBlockId(BlockId::Number(num)) => {
+                let num = num
+                    .as_number()
+                    .ok_or(anyhow!("invalid block number '{num}'"))?;
+                Ok(futures::stream::iter(num..num + 1))
+            }
+            BlockInterval::Range(range) => Ok(futures::stream::iter(range)),
+            _ => Err(anyhow!(
+                "could not create bounded stream from unbounded follow-from interval",
+            )),
+        }
+    }
+
+    /// Convert the block interval into an unbounded async stream of block
+    /// numbers. Query the blockchain node for the latest block number.
+    pub async fn into_unbounded_stream<ProviderT, TransportT>(
+        self,
+        provider: ProviderT,
+    ) -> Result<impl Stream<Item = Result<u64, anyhow::Error>>, anyhow::Error>
+    where
+        ProviderT: Provider<TransportT>,
+        TransportT: Transport + Clone,
+    {
+        match self {
+            BlockInterval::FollowFrom {
+                start_block,
+                block_time,
+            } => Ok(try_stream! {
+                let mut current = start_block;
+                 loop {
+                    let last_block_number = provider.get_block_number().await.map_err(|e: alloy::transports::RpcError<_>| {
```

Comment:
I'm surprised you can't elide the type. Also `.context` could work here

Reply:
Compiler insisted for some reason